crates/ide/src/goto_definition.rs

```diff
@@ -68,11 +78,49 @@ pub(crate) fn goto_definition(
     Some(RangeInfo::new(original_token.text_range(), nav_targets))
 }
 
+fn resolve_intra_doc_links(
+    position: FilePosition,
+    comment: ast::Comment,
+    file: SyntaxNode,
+    token: SyntaxToken,
+) -> SyntaxNode {
+    let comment_range = token.text_range();
+    if let Some(doc_comment) = comment.doc_comment() {
+        let mut min_distance = file.text_range().end();
+        let mut min_def_name = "".to_string();
+        let def_names = extract_definitions_from_markdown(doc_comment);
+        for def_name in def_names {
+            for (matched_index, _) in doc_comment.match_indices(&def_name) {
+                let matched_position = comment_range.start() + TextSize::from(matched_index as u32);
+                if let Some(distance) = position.offset.checked_sub(matched_position) {
+                    if let Some(_) = min_distance.checked_sub(distance) {
+                        min_distance = distance;
+                        min_def_name = def_name.clone();
+                    };
+                };
+            }
+        }
+        if let Some(candidate) = file.children().filter_map(ast::Fn::cast).find_map(|f| {
+            if let Some(name) = f.name() {
+                if let Some(ident_token) = name.ident_token() {
+                    if ident_token.to_string() == min_def_name {
+                        return Some(name);
+                    }
+                }
+            }
+            None
+        }) {
+            return candidate.syntax().clone();
+        }
+    };
+    file
+}
```

Comment:
Rather than implementing manual search based on syntax trees, this function should call into this thing, which already can resovle doc links:

https://github.com/rust-analyzer/rust-analyzer/blob/85a28751611e564eb4cdde28ca953407f551b696/crates/hir/src/attrs.rs#L20-L25

Reply:
I see. How to get the definition for `resolve_doc_path`? I can get the link (seems equivalent to the definition name string) and ns from markdown but no way to convert them to ast or hir structured data.

Reply:
See https://github.com/rust-analyzer/rust-analyzer/blob/1487f2f10efe2d0a54e8627692643f3a48ffdbff/crates/ide/src/runnables.rs#L149-L176